nibabel/gifti/gifti.py

```diff
@@ -280,21 +281,21 @@ def to_xml(self):
         # fix endianness to machine endianness
         self.endian = gifti_endian_codes.code[sys.byteorder]
         result = ""
-        result += self.to_xml_open()
+        result += self.to_xml_open().decode('utf-8')
         # write metadata
         if not self.meta is None:
-            result += self.meta.to_xml()
+            result += self.meta.to_xml().decode('utf-8')
         # write coord sys
         if not self.coordsys is None:
-            result += self.coordsys.to_xml()
+            result += self.coordsys.to_xml().decode('utf-8')
         # write data array depending on the encoding
         dt_kind = data_type_codes.dtype[self.datatype].kind
         result += data_tag(self.data,
                            gifti_encoding_codes.specs[self.encoding],
                            KIND2FMT[dt_kind],
-                           self.ind_ord)
-        result = result + self.to_xml_close()
-        return result
+                           self.ind_ord).decode('utf-8')
+        result = result + self.to_xml_close().decode('utf-8')
+        return result.encode('utf-8')
```

Comment:
This entire function is an encoded concatenation of decoded bytestrings. Can't all of these changes be dropped?

If you simply want to explicitly label this function as dealing in `bytes` and not `unicode`, you could use a `BytesIO` to accumulate:

    def to_xml(self):
        # fix endianness to machine endianness
        self.endian = gifti_endian_codes.code[sys.byteorder]
        result = BytesIO(self.to_xml_open())
        # write metadata
        if not self.meta is None:
            result.write(self.meta.to_xml())
        # write coord sys
        if not self.coordsys is None:
            result.write(self.coordsys.to_xml())
        # write data array depending on the encoding
        dt_kind = data_type_codes.dtype[self.datatype].kind
        result.write(data_tag(self.data,
                              gifti_encoding_codes.specs[self.encoding],
                              KIND2FMT[dt_kind],
                              self.ind_ord))
        result.write(self.to_xml_close())
        return result.getvalue()

While decoding and encoding will be slightly more expensive, writing to a BytesIO object will be slightly less.